ex1/exercise_1_x_mix_drix.py

```diff
@@ -67,40 +77,43 @@ def __init__(self, user_input_generator: Generator, ):
         self._board = XOBoard.empty_board()
 
     def change_player(self):
-        """
-        Changes the current player.
-        """
-        pass
+        if self._current_player == PlayerToken.X:
+            self._current_player = PlayerToken.O
+        else:
+            self._current_player = PlayerToken.X
 
     def current_player(self) -> PlayerToken:
-        """
-        Returns the token of the current player.
-        """
-        pass
+        return self._current_player
 
     def get_winner(self) -> PlayerToken:
-        """
-        Returns the token of the winning player if any of the player won.
-        Otherwise, returns None.
-        """
-        pass
+        if self._board.is_winning(X):
+            return X
+        elif self._board.is_winning(O):
+            return O
 
     def play_game(self, n_moves: int):
-        """
-        Plays n_moves turns of the game.
-        """
-        # Hint: take input from the user using next(self._user_input_generator)
-        pass
+        iterator = iter(self._user_input_generator)
+        for i in range(n_moves):
+            # Told for the user who is turn, collect a row and column from the user and try to put a play token there.
+            print(f"Now it's the turn of the player is play with {self._current_player.value}")
+            row, column = next(iterator)
+            self._board.play_move(row, column, self._current_player)
+            print(self._board)
+
+            # Check if the current player in won.
+            if self._board.is_winning(self._current_player) or " " not in str(self._board):
```

Comment:
Checking whether game is over using 
```python
" " not in str(self._board):
```
breaks the encapsulation of `XOBoard` class.
Do not break encapsulation - instead, create methods that inspect the inner data structure
and can do the required manipulation / calculation.